### DISCUSSION TITLE
Airflow Tasks being "Retried" instead of "Rescheduled"

### DISCUSSION THREAD

**Discussion by Vishal2696** [apache/airflow#27083] [Q&A] [answered]
**Airflow Version:** 2.2.3
**Metadata DB:** Postgres V11 (PAAS service by Azure - Single server)
**Deployment:** Kubernetes (webserver, worker, scheduler running as separate pods)
**Connection Pool:** Pgbouncer (Running as a separate pod)
**Executor:** Celery. (Redis is running as a separate pod)

**My Expectation:** 
I just need guidance from the community members on where the root-cause might be as I'm currently confused on what to look at. 

**My Problem:** 
My airflow setup was running stable for many months and then we had an Azure PostgresSQL outage one day. During the outage and also two days after the outage my Airflow tasks go into retry mode instead of being rescheduled (Rescheduled is the desired thing to happen according to our configuration). While we can easily conclude that this was due to the DB outage but the issue continued to happen 2 days post the DB recovered too. Another mysterious factor is post 2 days the issue automatically disappeared. 

**Errored out Task Logs From Airflow UI:** 
```
Log file does not exist: /opt/airflow/airflow/logs/mydagname/mytaskname/2022-10-15T04:00:00+00:00/10.log
*** Fetching from: http://worker:8793/log/mydagname/mytaskname/2022-10-15T04:00:00+00:00/10.log
*** Failed to fetch log file from worker. 404 Client Error: NOT FOUND for url: http://worker:8793/log/mydagname/mytaskname/2022-10-15T04:00:00+00:00/10.log
For more information check: https://httpstatuses.com/404
```

**Overview of what my task does:** 
It is a python operator task which checks an Azure eventhub for a particular msg. If the msg is found then it succeeds, if not it goes to "UP_FOR_RESCHEDULE" state and does the same check again after 5 minutes (time period set by me.)

**Temporary workaround I did:** 
Restarted the worker pod and the dags were rerun manually. This time they were rescheduled correctly instead of retried.

**Comment by potiuk** [comment, marked as answer]
When your task gets an error (for example DB failure) it will fail and retry rather than "reschedule". This is pretty much expected. I believe the "2 days after" were simply because some of your workers kept a connection that was cached in PGBouncer (possibly not shutdown properly0 and likely restarting it fixed the problem. Airflow does not have advanced mechanism to recover from all cases. This is kind of expected that if any component will go into outage related to network issues, you might need to restart all components. This depends on whether any connections that are kept are somehow cached/persisted and there is no easy way to solve it. Trying to recover from all possible issues which could happen in fully automated way is a huge investment (especially on trying to reproduce and test all scenarios), so the "reasonable" approach from Airlfow (since this is not a Highly Available critical system that must recover itself in all possible cases) is that in case you have problems like that, you should simply restart Airflow (including all its components). Airflow  is designed to recover after restart so you should be able to do it without any problems if you detect similar issue in the future.

**Reply by Vishal2696** [reply]
Thank you! This makes sense. Just wanted this to be confirmed by the community.
